### PR TITLE
Refactor/spring cleaning

### PR DESCRIPTION
**Pull request by Eikix**
Time spent on this PR: 2day

Resolves: #529 

- add more refactor with abigen-rs
- delete waiter
- add poll from starkli
- leaner codebase

# Pull Request type

Please check the type of change your PR introduces:

- [ ] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, renaming)
- [ ] Refactoring (no functional changes, no API changes)
- [ ] Build-related changes
- [ ] Documentation content changes
- [ ] Testing

# What is the new behavior?

# Does this introduce a breaking change?

- [ ] Yes
- [ ] No

### CODE REVIEW

File: crates/core/src/client/mod.rs [hunk 1]
```diff
@@ -68,38 +72,68 @@ impl<P: Provider + Send + Sync + 'static> KakarotClient<P> {
...
+            .eth_call(&origin, &to, &gas_limit, &gas_price, &value, &calldata.len().into(), &CairoArrayLegacy(calldata))
+            .block_id(starknet_block_id)
+            .call()
+            .await?;
+
+        if success == FieldElement::ZERO {
+            let revert_reason =
+                return_data.0.into_iter().filter_map(|x: FieldElement| u8::try_from(x).ok()).collect::<Vec<_>>();
+            let revert_reason = String::from_utf8(revert_reason).unwrap_or_else(|_| "Unknown".into());
+            return Err(EthApiError::Other(anyhow!("Revert reason: {}", revert_reason)));
```
Comment (Eikix) [inline crates/core/src/client/mod.rs:131]: Maybe here we should not return an Error, semantically, maybe semantically it isn't an error

File: crates/core/src/client/mod.rs [hunk 2]
```diff
@@ -68,38 +72,68 @@ impl<P: Provider + Send + Sync + 'static> KakarotClient<P> {
...
+            .reader
+            .eth_call(&origin, &to, &gas_limit, &gas_price, &value, &calldata.len().into(), &CairoArrayLegacy(calldata))
+            .block_id(starknet_block_id)
+            .call()
+            .await?;
+
+        if success == FieldElement::ZERO {
+            let revert_reason =
+                return_data.0.into_iter().filter_map(|x: FieldElement| u8::try_from(x).ok()).collect::<Vec<_>>();
+            let revert_reason = String::from_utf8(revert_reason).unwrap_or_else(|_| "Unknown".into());
```
Comment (greged93) [inline crates/core/src/client/mod.rs:130]:
```suggestion
            let revert_reason =
                return_data.0.into_iter().filter_map(|x| u8::try_from(x).ok()).map(|x| x as char).collect::<String>();
```

File: crates/core/src/client/mod.rs (hunk 1 above)
Comment (greged93) [reply crates/core/src/client/mod.rs:131]: I agree, would be inclined to return the revert_reason converted to bytes?

Comment (greged93) [reply crates/core/src/client/mod.rs:131]: I checked `reth`, they return an error if the execution failed, we could add `ExecutionError` to the `EthApiError`
https://github.com/paradigmxyz/reth/blob/main/crates/rpc/rpc/src/eth/api/call.rs#L69C12-L69C12

File: crates/core/src/contracts/erc20/ethereum_erc20.rs [hunk 3]
```diff
@@ -24,30 +29,57 @@ abigen!(
...
+impl<P: Provider + Send + Sync + 'static> EthereumErc20<P> {
+    pub const fn new(address: FieldElement, provider: Arc<P>, kakarot_address: FieldElement) -> Self {
+        Self { address, provider, kakarot_address }
     }
 
     pub async fn balance_of(self, evm_address: Address, block_id: BlockId) -> Result<U256, EthApiError> {
         // Prepare the calldata for the bytecode function call
         let calldata = IERC20Calls::BalanceOf(BalanceOfCall { account: evm_address }).encode();
-        let calldata = calldata.into_iter().map(FieldElement::from).collect();
+        let calldata: Vec<FieldElement> = calldata.into_iter().map(FieldElement::from).collect();
```
Comment (greged93) [inline crates/core/src/contracts/erc20/ethereum_erc20.rs:46]: is this needed?

File: crates/core/src/contracts/erc20/ethereum_erc20.rs [hunk 4]
```diff
@@ -24,30 +29,57 @@ abigen!(
...
+                &CairoArrayLegacy(calldata),
+            )
+            .block_id(block_id)
+            .call()
+            .await?;
+
+        if success == FieldElement::ZERO {
+            let revert_reason =
+                return_data.0.into_iter().filter_map(|x: FieldElement| u8::try_from(x).ok()).collect::<Vec<_>>();
+            let revert_reason = String::from_utf8(revert_reason).unwrap_or_else(|_| "Unknown".into());
```
Comment (greged93) [inline crates/core/src/contracts/erc20/ethereum_erc20.rs:76]: same change as above

File: crates/test-utils/src/execution/mod.rs [hunk 5]
```diff
@@ -1,2 +1,49 @@
...
+                }
+                ExecutionResult::Reverted { reason } => {
+                    return Err(anyhow::anyhow!("transaction reverted: {}", reason));
+                }
+            },
+            Err(ProviderError::StarknetError(StarknetErrorWithMessage {
+                code: MaybeUnknownErrorCode::Known(StarknetError::TransactionHashNotFound),
+                ..
+            })) => {
+                eprintln!("Transaction not confirmed yet...");
```
Comment (greged93) [inline crates/test-utils/src/execution/mod.rs:33]: can we use the tracing crate here?

File: crates/test-utils/src/execution/mod.rs [hunk 6]
```diff
@@ -1,2 +1,49 @@
...
+            }
+            // Some nodes are still serving error code `25` for tx hash not found. This is
+            // technically a bug on the node's side, but we maximize compatibility here by also
+            // accepting it.
+            Err(ProviderError::StarknetError(StarknetErrorWithMessage {
+                code: MaybeUnknownErrorCode::Known(StarknetError::InvalidTransactionHash),
+                ..
+            })) => {
+                eprintln!("Transaction not confirmed yet...");
```
Comment (greged93) [inline crates/test-utils/src/execution/mod.rs:42]: same

Comment (Eikix) [reply crates/test-utils/src/execution/mod.rs:42]: i don't know what you mean

File: crates/core/src/contracts/erc20/ethereum_erc20.rs (hunk 3 above)
Comment (Eikix) [reply crates/core/src/contracts/erc20/ethereum_erc20.rs:46]: afaik we need a vec of felts to pass to the contract call

Comment (greged93) [reply crates/core/src/contracts/erc20/ethereum_erc20.rs:46]: ah yeah but do you need to explicitly type it? Compiler can't derive the type?

Comment (Eikix) [reply crates/core/src/contracts/erc20/ethereum_erc20.rs:46]: can try

**Pull request merged**